File: Best_Time_to_Buy_and_Sell_Stock_III.java
/*123. Best Time to Buy and Sell Stock III Add to List
Description  Submission  Solutions
Total Accepted: 79172
Total Submissions: 277007
Difficulty: Hard
Contributors: Admin
Say you have an array for which the ith element is the price of a given stock on day i.

Design an algorithm to find the maximum profit. You may complete at most two transactions.

Note:
You may not engage in multiple transactions at the same time (ie, you must sell the stock before you buy again).

Hide Tags Array Dynamic Programming
Hide Similar Problems (E) Best Time to Buy and Sell Stock (E) Best Time to Buy and Sell Stock II (H) Best Time to Buy and Sell Stock IV
 */






/*
assume only have 0 money at first (4 ms/ 198 test)
"sold2" --- the maximum if we've just sold 2nd stock so far
"buy2" --- the maximum if we've jest buy 2nd stock
"sold1" --- the maximum if we've just sold 1st stock so far
"buy1" --- the maximum if we've jest buy 1st stock
 */

////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public int maxProfit(int[] prices) {
        if(prices==null || prices.length==0) return 0;
        int buy1 = Integer.MIN_VALUE;
        int buy2 = Integer.MIN_VALUE;
        int sold1 = 0;
        int sold2 = 0;
        for(int price: prices){
            sold2 = Math.max(sold2, buy2 + price);
            buy2 = Math.max(buy2, sold1 - price);
            sold1 = Math.max(sold1, buy1 + price);
            buy1 = Math.max(buy1, -price);
        }
        //sold2 always higher than sold1
        return sold2;
    }
}

File: Best_Time_to_Buy_and_Sell_Stock_III.swift
/*
You are given an array prices where prices[i] is the price of a given stock on the ith day.

Find the maximum profit you can achieve. You may complete at most two transactions.

Note: You may not engage in multiple transactions simultaneously (i.e., you must sell the stock before you buy again).



Example 1:

Input: prices = [3,3,5,0,0,3,1,4]
Output: 6
Explanation: Buy on day 4 (price = 0) and sell on day 6 (price = 3), profit = 3-0 = 3.
Then buy on day 7 (price = 1) and sell on day 8 (price = 4), profit = 4-1 = 3.
Example 2:

Input: prices = [1,2,3,4,5]
Output: 4
Explanation: Buy on day 1 (price = 1) and sell on day 5 (price = 5), profit = 5-1 = 4.
Note that you cannot buy on day 1, buy on day 2 and sell them later, as you are engaging multiple transactions at the same time. You must sell before buying again.
Example 3:

Input: prices = [7,6,4,3,1]
Output: 0
Explanation: In this case, no transaction is done, i.e. max profit = 0.
Example 4:

Input: prices = [1]
Output: 0


Constraints:

1 <= prices.length <= 105
0 <= prices[i] <= 105
*/

/*
Solution 1: DP
"sold2" --- the maximum if we've just sold 2nd stock so far
"buy2" --- the maximum if we've jest buy 2nd stock
"sold1" --- the maximum if we've just sold 1st stock so far
"buy1" --- the maximum if we've jest buy 1st stock

return s2 would enough
sold always higher than buy
s2 = max(s2, b2 + p)
b2 = max(b2, s1 - p)
==> s2 = max(s2, s1), s2 always the higher one

Time Complexity: O(n)
Space Complexity: O(1)
*/
class Solution {
    func maxProfit(_ prices: [Int]) -> Int {
        let n = prices.count

        var buy1 = Int.min
        var buy2 = Int.min
        var sell1 = 0
        var sell2 = 0

        for price in prices {
            sell2 = max(sell2, buy2 + price)
            buy2 = max(buy2, sell1 - price)

            sell1 = max(sell1, buy1 + price)
            buy1 = max(buy1, -price)
        }

        // sell2 always higher than sell1
        return sell2
    }
}
